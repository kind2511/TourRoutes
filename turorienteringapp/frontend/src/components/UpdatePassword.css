.update-password-container {
    width: 500px;
    padding: 20px;
    border-radius: 10px;
    box-shadow: 0px 0px 15px rgba(19, 182, 38, 0.937);
    background-color: #d5d2d2;
    position: fixed;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    display: flex;
    flex-direction: column;
    align-items: center;
}

.update-password-container input {
    width: 80%;  
    padding: 8px;  
    margin-bottom: 20px;  
    border: 1px solid #ddd;
    border-radius: 5px;
    font-size: 14px;  
}

.update-password-container p {
    margin-bottom: 20px;
    font-size: 16px;
    font-weight: bold;
}

.update-password-buttons {
    display: flex;
    justify-content: space-between;
    width: 100%;
}

.update-password-container .update-password-btn,
.update-password-container .cancel-btn {
    padding: 10px 20px;
    width: 48%;  
    margin: 0 1%; /* Margin between buttons */
    border: none;
    border-radius: 5px;
    cursor: pointer;
    font-size: 16px;
    color: white;
}

/* Styles for buttons inside the update password container */
.update-password-container .update-password-btn {
    background-color: #E53E3E; /* Red button for 'Change Password' */
   
  }

.update-password-container .cancel-btn {
    background-color: #4CAF50;  
     
  }

/* Hover effects */
.update-password-container .update-password-btn:hover,
.update-password-container .cancel-btn:hover {
  background-color: #61dafb;  
  color: #333;  
}

.update-password-error {
    color: #E53E3E;  
    margin-bottom: 10px;  
  }
  
  .update-password-success {
    font-size: 16px;
    font-weight: bold;
    color: #E53E3E;
  }