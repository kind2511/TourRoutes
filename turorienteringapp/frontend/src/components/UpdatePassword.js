import React, { useState } from "react";
import "./UpdatePassword.css";

const UpdatePassword = ({ close }) => {
  const [password, setPassword] = useState('');
  const [confirmPassword, setConfirmPassword] = useState('');
  const [error, setError] = useState('');

  //------------------------------------------------------->

  // Function to validate password length
  const validatePasswordLength = (password) => {
    if (password.length < 8) {
      setError("Password must be at least 8 characters.");
    } else {
      setError(""); // Clear error if length requirement  meet
    }
  };

  
  //------------------------------------------------------->

  // Function to validate that passwords match
  const validatePasswordsMatch = (password, confirmPassword) => {
    // Only check if there is no length error and both passwords have been entered
    if (!error && password && confirmPassword && password !== confirmPassword) {
      setError("Passwords do not match.");
    } else {
      setError(""); // Clear error if passwords match or one of them is still empty
    }
  };

  
  //------------------------------------------------------->

  //validate password length on every change
  const handlePasswordChange = (e) => {
    const newPassword = e.target.value;
    setPassword(newPassword);
    validatePasswordLength(newPassword);
    // If confirmPassword is already entered, validate for match
    if (confirmPassword) {
      validatePasswordsMatch(newPassword, confirmPassword);
    }
  };

  
  //------------------------------------------------------->

  //validate passwords match on every change
  const handleConfirmPasswordChange = (e) => {
    const newConfirmPassword = e.target.value;
    setConfirmPassword(newConfirmPassword);
    // Validate for match only if write in input feild
    if (password) {
      validatePasswordsMatch(password, newConfirmPassword);
    }
  };

  
  //------------------------------------------------------->

  //Logic to handle update the password 

  const handleUpdate = async () => {
    // If there's an error, don't proceed  the API call
    if (error) return;

    // Proceed with the API call if validations pass
    try {
      const response = await fetch('http://localhost:8000/api/v1/users/updateMyPassword', {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${localStorage.getItem('token')}`,
        },
        body: JSON.stringify({ password, confirmPassword }),
      });

      const data = await response.json();
      if (response.ok) {
        // Password updated successfully
        close(); // Close the update password dialog
      } else {
        // Handle API errors_ backend requirements
        setError(data.message || "Failed to update password.");
      }
    } catch (err) {
      setError("An error occurred while updating the password.");
      console.error("Error updating password:", err);
    }
  };
  
  //------------------------------------------------------->

  //Rendring
  return (
    <div className="update-password-container">
      <h1>Update Password</h1>
      {error && <p className="update-password-error">{error}</p>}
      <input
        type="password"
        placeholder="New Password"
        value={password}
        onChange={handlePasswordChange}
      />
      <input
        type="password"
        placeholder="Confirm New Password"
        value={confirmPassword}
        onChange={handleConfirmPasswordChange}
      />
      <div className="update-password-buttons">
        <button className="update-password-btn" onClick={handleUpdate}>Update Password</button>
        <button className="cancel-btn" onClick={close}>Cancel</button>
      </div>
    </div>
  );
};

export default UpdatePassword;