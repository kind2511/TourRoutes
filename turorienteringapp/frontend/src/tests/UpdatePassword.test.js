import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import UpdatePassword from '../components/UpdatePassword';

test('UpdatePassword component renders without crashing', () => {
  render(<UpdatePassword close={() => { }} />);
});

//---------------------------------------------->

// Snapshot test for the update pasword component
test('UpdatePassword component snapshot', () => {
  const { asFragment } = render(
    <BrowserRouter>
      <UpdatePassword />
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
