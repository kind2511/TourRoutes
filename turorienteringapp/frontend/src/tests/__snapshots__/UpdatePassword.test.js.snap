// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdatePassword component snapshot 1`] = `
<DocumentFragment>
  <div
    class="update-password-container"
  >
    <h1>
      Update Password
    </h1>
    <input
      placeholder="New Password"
      type="password"
      value=""
    />
    <input
      placeholder="Confirm New Password"
      type="password"
      value=""
    />
    <div
      class="update-password-buttons"
    >
      <button
        class="update-password-btn"
      >
        Update Password
      </button>
      <button
        class="cancel-btn"
      >
        Cancel
      </button>
    </div>
  </div>
</DocumentFragment>
`;
